File: 34. Find First and Last Position of Element in Sorted Array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if target not in nums:
            return [-1, -1]
        left = self.helper(nums, target, True)
        if left >= len(nums) or nums[left] != target:
            return [-1, -1]
        return [left, self.helper(nums, target, False)-1]
        
    def helper(self, nums, target, isleft):
        '''
        Careful about leftmost search and binary search
        '''
        left, right = 0, len(nums)
        while left < right:
            mid = (right-left)//2 + left
            if nums[mid] > target or (isleft and nums[mid] == target):
                right = mid
            else:
                left = mid + 1
        return left

File: 34. Find First and Last Position of Element in Sorted Array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        /**
         * Leftmost search and Rightmost search
         * Time O(logn)
         * Space O(1)
         */
        if (nums == null || nums.length == 0) {
            return new int[] {-1,-1};
        }
        int[] res = new int[2];
        int lo = 0;
        int hi = nums.length;
        while (lo < hi) {
            int mid = (hi - lo)/2 + lo;
            if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        if (lo >= nums.length || nums[lo] != target) {
            return new int[] {-1,-1};
        }
        res[0] = lo;
        lo = 0;
        hi = nums.length;
        while (lo < hi) {
            int mid = (hi - lo)/2 + lo;
            if (nums[mid] <= target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        res[1] = lo - 1;
        return res;
    }
}